# Part 1. Equation of a Slime

How many late days are you using for this assignment?

In [25]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score
import matplotlib.pyplot as plt

## 1. Loading the dataset

In [26]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

science_data =pd.read_csv('science_data_large.csv')
display(science_data.head(15))

display(science_data.info())
display(science_data.describe())

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


None

,Temperature °C,Mols KCL,Size nm^3
count,1000.000000,1000.000000,1.000000e+03
mean,500.500000,471.530000,5.086111e+05
std,288.819436,288.482872,4.474838e+05
min,1.000000,1.000000,1.611429e+01
25%,250.750000,226.750000,1.298267e+05
50%,500.500000,459.500000,3.827182e+05
75%,750.250000,710.250000,7.603211e+05
max,1000.000000,1000.000000,1.972127e+06


## 2. Splitting the dataset

In [27]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = science_data[['Temperature °C', 'Mols KCL']]  
y = science_data['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.1, #(90% train, 10% test)
    random_state=42
)

print("Training features shape:", X_train.shape)
print("Testing features shape:", X_test.shape)
print("Training labels shape:", y_train.shape)
print("Testing labels shape:", y_test.shape)

display(X_train.head())


Training features shape: (900, 2)
Testing features shape: (100, 2)
Training labels shape: (900,)
Testing labels shape: (100,)


,Temperature °C,Mols KCL
716,95,499
351,243,384
936,47,211
256,495,497
635,82,421


## 3. Perform a Linear Regression

In [28]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = [[200, 200]]  #temperature: 200, Mols KCL: 200
predicted_size = model.predict(sample)
print(f"Predicted size for sample {sample}: {predicted_size[0]:.2f} nm^3")
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(f"\nModel R-squared score: {score:.4f}")
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coef_temp = model.coef_[0]
coef_kcl = model.coef_[1]
intercept = model.intercept_

print("\nModel coefficients:")
print(f"Temperature coefficient: {coef_temp:.2f}")
print(f"Mols KCL coefficient: {coef_kcl:.2f}")
print(f"Intercept: {intercept:.2f}")
#Using LaTeX
print("\nRegression equation:")
print(r"h(x) = {:.2f} + {:.2f} \times \text{{Temperature}} + {:.2f} \times \text{{Mols KCL}}".format(
    intercept, coef_temp, coef_kcl))


Predicted size for sample [[200, 200]]: -29623.18 nm^3

Model R-squared score: 0.8552

Model coefficients:
Temperature coefficient: 866.15
Mols KCL coefficient: 1032.70
Intercept: -409391.48

Regression equation:
h(x) = -409391.48 + 866.15 \times \text{Temperature} + 1032.70 \times \text{Mols KCL}


/opt/anaconda3/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)

Size=−409391.48+866.15×Temperature+1032.70×Mols KCL

Report on score and explain meaning:
The score is 0.8552, which means that there is a 85.52% variability in slime size and consequently a pretty strong relationship between temperature, KCL and slime growth

## 4. Use Cross Validation

In [29]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
from sklearn.model_selection import cross_val_score, KFold

kfold = KFold(n_splits=5, shuffle=True, random_state=42)

cv_scores = cross_val_score(LinearRegression(), X, y, 
                          cv=kfold,  
                          scoring='r2') 

print("Cross-validation R-squared scores:", cv_scores)
print(f"Mean R-squared: {cv_scores.mean():.4f}")
print(f"Standard deviation: {cv_scores.std():.4f}")
# Report on their finding and their significance

Cross-validation R-squared scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean R-squared: 0.8597
Standard deviation: 0.0184


Write findings here:

The Mean R-squared of this test was 0.8597, with a standard deviation of 0.0184

## 5. Using Polynomial Regression

In [30]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create polynomial features of degree 2
poly = PolynomialFeatures(degree=2)
X_poly_train = poly.fit_transform(X_train)
X_poly_test = poly.transform(X_test)

# Train a linear regression model on the polynomial features
model_poly = LinearRegression()
model_poly.fit(X_poly_train, y_train)

# Perform k-fold cross-validation
cv_scores = cross_val_score(model_poly, X_poly_train, y_train, cv=5)  # 5-fold CV
print(f"Cross-validation scores: {cv_scores}")
print(f"Mean R-squared score from CV: {cv_scores.mean():.4f}")

# Extract the coefficients and intercept
coefficients = model_poly.coef_
intercept = model_poly.intercept_

# Output the polynomial regression equation
print("\nPolynomial Regression Equation:")
terms = [f"{coefficients[i]:.2f} * {term}" for i, term in enumerate(poly.get_feature_names_out())]
equation = " + ".join(terms)
print(f"h(x) = {intercept:.2f} + {equation}")
# Report on the metrics and output the resultant equation as you did in Part 3.

Cross-validation scores: [1. 1. 1. 1. 1.]
Mean R-squared score from CV: 1.0000

Polynomial Regression Equation:
h(x) = 0.00 + 0.00 * 1 + 12.00 * Temperature °C + -0.00 * Mols KCL + -0.00 * Temperature °C^2 + 2.00 * Temperature °C Mols KCL + 0.03 * Mols KCL^2


The Cross-validation scores: [1. 1. 1. 1. 1.]
The Mean R-squared score from CV: 1.0000



Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Polynomial Regression Equation:
h(x) = 0.00 + 0.00 * 1 + 12.00 * Temperature °C + -0.00 * Mols KCL + -0.00 * Temperature °C^2 + 2.00 * Temperature °C Mols KCL + 0.03 * Mols KCL^2

Report on the score and interpret:

Since the R^2 score is 0.8552, the regression model is effective.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [31]:
# Load the dataset. Then train and evaluate the classification models.
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, confusion_matrix


df = pd.read_csv('ckd_feature_subset.csv')
print(df.head())
print(df.info())
print(df.describe())

print("Missing values:\n", df.isnull().sum())

X = df.drop('Target_ckd', axis=1)
y = df['Target_ckd']
y = np.where(y == 'yes', 1, np.where(y == 'no', 0, y))

models = {
    'Logistic Regression': LogisticRegression(max_iter=200),
    'Support Vector Machines': SVC(),
    'k-Nearest Neighbors': KNeighborsClassifier(),
    'Random Forest': RandomForestClassifier(random_state=42),
    'Neural Network': MLPClassifier(max_iter=200)
}

results = {}
for model_name, model in models.items():
    cv_scores = cross_val_score(model, X, y, cv=5)
    results[model_name] = {
        'Mean Accuracy': np.mean(cv_scores),
        'Std Dev': np.std(cv_scores)
    }

results_df = pd.DataFrame(results).T
print("\nCross-Validation Results:")
print(results_df)

model_results = {}
for model_name, model in models.items():
    model.fit(X, y)
    y_pred = model.predict(X)
    accuracy = accuracy_score(y, y_pred)
    report = classification_report(y, y_pred)
    confusion = confusion_matrix(y, y_pred)
    model_results[model_name] = {
        'Accuracy': accuracy,
        'Classification Report': report,
        'Confusion Matrix': confusion
    }

for model_name, results in model_results.items():
    print(f"\nModel: {model_name}")
    print(f"Accuracy: {results['Accuracy']:.2f}")
    print("Classification Report:\n", results['Classification Report'])
    print("Confusion Matrix:\n", results['Confusion Matrix'])

        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   age         153 non-null    float64
 1   bp          153 non-null    float64
 2   wbcc        153 non-null    float64
 3   appet_poor  153 non-null    int64  
 4   appet_good  153 non-null    int64  
 5   rbcc        153 non-null    float64
 6   Target_ckd  153 non-null    int64  
dtypes: float64(4), int64(3)
memory usage: 8.5 KB
None
    

/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_


Cross-Validation Results:
                         Mean Accuracy   Std Dev
Logistic Regression           0.856129  0.054236
Support Vector Machines       0.928172  0.059993
k-Nearest Neighbors           0.928172  0.031893
Random Forest                 0.993333  0.013333
Neural Network                0.882151  0.049926

Model: Logistic Regression
Accuracy: 0.87
Classification Report:
               precision    recall  f1-score   support

           0       0.85      1.00      0.92       111
           1       1.00      0.52      0.69        42

    accuracy                           0.87       153
   macro avg       0.92      0.76      0.80       153
weighted avg       0.89      0.87      0.85       153

Confusion Matrix:
 [[111   0]
 [ 20  22]]

Model: Support Vector Machines
Accuracy: 0.95
Classification Report:
               precision    recall  f1-score   support

           0       0.94      1.00      0.97       111
           1       1.00      0.83      0.91        42

    accu

/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments

For CKD cases: The model made 10 correct predictions (TP) out of 10 actual CKD cases (meaning it didn't misclassify any CKD patients). And for non-CKD cases: The model made 21 correct predictions (TN) out of 21 actual non-CKD cases (again, no misclassifications). However, with the high accuracy noted, it is possible that the models were overfitting.

In [32]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

df = pd.read_csv('ckd_feature_subset.csv')
print(df.head())
print(df.info())
print(df.describe())

print("Missing values:\n", df.isnull().sum())

X = df.drop('Target_ckd', axis=1)
y = df['Target_ckd']
y = np.where(y == 'yes', 1, np.where(y == 'no', 0, y))

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#Different Neural Networks
nn_configurations = {
    'Config 1: Hidden Layer (5,)': MLPClassifier(hidden_layer_sizes=(5,), max_iter=200, random_state=42),
    'Config 2: Hidden Layer (10,)': MLPClassifier(hidden_layer_sizes=(10,), max_iter=200, random_state=42),
    'Config 3: Two Hidden Layers (10, 5)': MLPClassifier(hidden_layer_sizes=(10, 5), max_iter=200, random_state=42),
    'Config 4: Learning Rate (0.01), Hidden Layer (10,)': MLPClassifier(hidden_layer_sizes=(10,), max_iter=200, learning_rate_init=0.01, random_state=42),
    'Config 5: Learning Rate (0.001), Hidden Layer (20,)': MLPClassifier(hidden_layer_sizes=(20,), max_iter=200, learning_rate_init=0.001, random_state=42)
}

nn_results = {}
for config_name, nn_model in nn_configurations.items():
    cv_scores = cross_val_score(nn_model, X, y, cv=5)
    nn_results[config_name] = {
        'Mean Accuracy': np.mean(cv_scores),
        'Std Dev': np.std(cv_scores)
    }

nn_results_df = pd.DataFrame(nn_results).T
print("\nNeural Network Configurations Results:")
print(nn_results_df)

best_config_name = 'Config 4: Learning Rate (0.01), Hidden Layer (10,)'
best_nn_model = nn_configurations[best_config_name]

best_nn_model.fit(X_train, y_train)
y_pred = best_nn_model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
confusion = confusion_matrix(y_test, y_pred)

print(f"\nBest Config Model: {best_config_name}")
print(f"Accuracy: {accuracy:.2f}")
print("Classification Report:\n", report)
print("Confusion Matrix:\n", confusion)

/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


        age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0  0.688312  0.333333  0.000000           1           0  0.000000           1
1  0.545455  0.333333  0.128319           1           0  0.305085           1
2  0.714286  0.500000  0.238938           1           0  0.186441           1
3  0.688312  0.333333  0.283186           0           1  0.338983           1
4  0.441558  0.333333  0.221239           1           0  0.220339           1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   age         153 non-null    float64
 1   bp          153 non-null    float64
 2   wbcc        153 non-null    float64
 3   appet_poor  153 non-null    int64  
 4   appet_good  153 non-null    int64  
 5   rbcc        153 non-null    float64
 6   Target_ckd  153 non-null    int64  
dtypes: float64(4), int64(3)
memory usage: 8.5 KB
None
    

/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_


Neural Network Configurations Results:
                                                    Mean Accuracy   Std Dev
Config 1: Hidden Layer (5,)                              0.725591  0.013368
Config 2: Hidden Layer (10,)                             0.849677  0.054005
Config 3: Two Hidden Layers (10, 5)                      0.849677  0.054005
Config 4: Learning Rate (0.01), Hidden Layer (10,)       0.941290  0.024040
Config 5: Learning Rate (0.001), Hidden Layer (...       0.849677  0.054005

Best Config Model: Config 4: Learning Rate (0.01), Hidden Layer (10,)
Accuracy: 0.90
Classification Report:
               precision    recall  f1-score   support

           0       0.95      0.90      0.93        21
           1       0.82      0.90      0.86        10

    accuracy                           0.90        31
   macro avg       0.88      0.90      0.89        31
weighted avg       0.91      0.90      0.90        31

Confusion Matrix:
 [[19  2]
 [ 1  9]]


/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/opt/anaconda3/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

It seems that the neural network with a learning rate of 0.01 and one hidden layer of 10 neurons outperformed the others in terms of both accuracy and stability (low standard deviation). The careful tuning of learning rates along with the model architecture are really important for achieving high performance, and future experiments should focus on hyperparameter tuning or feature engineering for the most improvements.